src/test/java/org/apache/sysds/test/functions/federated/codegen/FederatedRowwiseTmplTest.java

```diff
@@ -0,0 +1,181 @@
+			{20, 1500, 8},
+			{21, 1500, 8},
+			{25, 600, 10},
+			{31, 150, 10},
+			{40, 300, 20},
+			{45, 1500, 100},
+			{50, 376, 4},
+		});
+	}
+
+	@BeforeClass
+	public static void init() {
+		TestUtils.clearDirectory(TEST_DATA_DIR + TEST_CLASS_DIR);
+	}
+
+	// @Test
+	// public void federatedCodegenRowwiseSingleNode() {
+	// 	testFederatedCodegenRowwise(ExecMode.SINGLE_NODE);
+	// }
+	// 
+	// @Test
+	// public void federatedCodegenRowwiseSpark() {
+	// 	testFederatedCodegenRowwise(ExecMode.SPARK);
+	// }
+	
+	@Test
+	public void federatedCodegenCellwiseHybrid() {
+		testFederatedCodegenRowwise(ExecMode.HYBRID);
+	}
+
+	private void testFederatedCodegenRowwise(ExecMode exec_mode) {
+		// store the previous platform config to restore it after the test
+		ExecMode platform_old = setExecMode(exec_mode);
+
+		getAndLoadTestConfiguration(TEST_NAME);
+		String HOME = SCRIPT_DIR + TEST_DIR;
+
+		int fed_rows = rows / 2;
+		int fed_cols = cols;
+
+		// generate dataset
+		// matrix handled by two federated workers
+		double[][] X1 = getRandomMatrix(fed_rows, fed_cols, 0, 1, 1, 3);
+		double[][] X2 = getRandomMatrix(fed_rows, fed_cols, 0, 1, 1, 7);
+
+		writeInputMatrixWithMTD("X1", X1, false, new MatrixCharacteristics(fed_rows, fed_cols, BLOCKSIZE, fed_rows * fed_cols));
+		writeInputMatrixWithMTD("X2", X2, false, new MatrixCharacteristics(fed_rows, fed_cols, BLOCKSIZE, fed_rows * fed_cols));
+
+		// empty script name because we don't execute any script, just start the worker
+		fullDMLScriptName = "";
+		int port1 = getRandomAvailablePort();
+		int port2 = getRandomAvailablePort();
+		Thread thread1 = startLocalFedWorkerThread(port1, FED_WORKER_WAIT_S);
+		Thread thread2 = startLocalFedWorkerThread(port2);
+
+		getAndLoadTestConfiguration(TEST_NAME);
+
+		// Run reference dml script with normal matrix
+		fullDMLScriptName = HOME + TEST_NAME + "Reference.dml";
+		programArgs = new String[] {"-stats", "-nvargs",
+			"in_X1=" + input("X1"), "in_X2=" + input("X2"),
+			"in_test_num=" + Integer.toString(test_num),
+			"out_Z=" + expected(OUTPUT_NAME)};
+		runTest(true, false, null, -1);
+
+		// Run actual dml script with federated matrix
+		fullDMLScriptName = HOME + TEST_NAME + ".dml";
+		programArgs = new String[] {"-stats", "-nvargs",
+			"in_X1=" + TestUtils.federatedAddress(port1, input("X1")),
+			"in_X2=" + TestUtils.federatedAddress(port2, input("X2")),
+			"in_test_num=" + Integer.toString(test_num),
+			"rows=" + rows, "cols=" + cols,
+			"out_Z=" + output(OUTPUT_NAME)};
+		runTest(true, false, null, -1);
+
+		// compare the results via files
+		HashMap<CellIndex, Double> refResults  = readDMLMatrixFromExpectedDir(OUTPUT_NAME);
+		HashMap<CellIndex, Double> fedResults = readDMLMatrixFromOutputDir(OUTPUT_NAME);
+		TestUtils.compareMatrices(fedResults, refResults, TOLERANCE, "Fed", "Ref");
+
+		TestUtils.shutdownThreads(thread1, thread2);
+
+		// check for federated operations
+		Assert.assertTrue(heavyHittersContainsSubString("fed_spoofRA"));
```

Comment:
just verifying that it calls the spoof operation is fine, but i would like some verification that the result is correct as well.
If this is not possible in all your tests cases i suggest adding one where you have a known result, that you then verify calculates correctly.

Reply:
nvm did not look correctly at line 157